### PR TITLE
[inductor][cpu] Stop lowering div to reciprocal multiplication to preserve precision when the divisor is a scalar and device is on cpu

### LINKED ISSUE #157959: Torch compile don't work correctly with divide by scalar
### 🐛 Describe the bug

When you divide by scalar you get diffrent results in torch.compile. One solution is to put the scalar into tensor with nn.Parameters to solve the issue.

```python
import torch
from torch import nn

class Foo(nn.Module):

    def __init__(
        self,
        use_parameter: bool
    ) -> None:
        super().__init__()
        self.b = 101
        if use_parameter:
            self.b = nn.Parameter(torch.Tensor([self.b]), requires_grad=False)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # return x + self.b
        # return x - self.b
        return x / self.b
        # return x * self.b
    
torch.manual_seed(42)
x = torch.rand((5, 5))
expected = Foo(False)(x)

models = [
    Foo(False),
    Foo(True),
    torch.compile(Foo(False), fullgraph=True),
    torch.compile(Foo(True), fullgraph=True),
]

for m in models:
    print((m(x) - expected).sum())
```

Results i got are:
```
tensor(0.)
tensor(0.)
tensor(-2.6776e-09)
tensor(0.)
```

### Versions

gpytorch==1.14
pytorch-lightning==2.5.2
torch==2.7.1
torchinfo==1.8.0
torchmetrics==1.3.1

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### PR DESCRIPTION
**Pull request by thenumberouscode**
## mini repro from issue
```c++
import torch
from torch import nn

class Foo(nn.Module):

    def __init__(
        self,
        use_parameter: bool
    ) -> None:
        super().__init__()
        self.b = 101
        if use_parameter:
            self.b = nn.Parameter(torch.Tensor([self.b]), requires_grad=False)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # return x + self.b
        # return x - self.b
        return x / self.b
        # return x * self.b
    
torch.manual_seed(42)
x = torch.rand((5, 5))
expected = Foo(False)(x)

models = [
    Foo(False),
    Foo(True),
    torch.compile(Foo(False), fullgraph=True),
    torch.compile(Foo(True), fullgraph=True),
]

for m in models:
    print((m(x) - expected).sum())
```

all outputs equal zero except the result of  torch.compile(Foo(False), fullgraph=True)

## summary:
when divisor is a scalar, inductor will lower div to mul the scalar's reciprocal.
this could lead precision lost in c++ kernel. but not in triton kernel
## why:
Generated C++ kernel; thanks to @xmfan for supplying the code.
```c++
#include <torch/csrc/inductor/cpp_prefix.h>
extern "C"  void kernel(const float* in_ptr0,
                       float* out_ptr0)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(25L); x0+=static_cast<int64_t>(16L))
        {
            {
                if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(16L)))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                    auto tmp1 = static_cast<float>(0.009900990099009901);
                    auto tmp2 = at::vec::Vectorized<float>(tmp1);
                    auto tmp3 = tmp0 * tmp2;
                    tmp3.store(out_ptr0 + static_cast<int64_t>(x0));
                }
                if(C10_UNLIKELY(x0 >= static_cast<int64_t>(16L) && x0 < static_cast<int64_t>(25L)))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(9L));
                    auto tmp1 = static_cast<float>(0.009900990099009901);
                    auto tmp2 = at::vec::Vectorized<float>(tmp1);
                    auto tmp3 = tmp0 * tmp2;
                    tmp3.store(out_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(9L));
                }
            }
        }
    }
}
```
The float type in C typically has 6 to 7 significant digits, while the double type has 15 to 16 significant digits.
```c++
#include <iostream>
#include <iomanip>

int main() {
 auto tmp1 = static_cast<float>(0.009900990099009901);
 auto tmp2 = static_cast<double>(0.009900990099009901);
 std::cout << std::setprecision(20) << "tmp1 = " << tmp1 << std::endl;
 std::cout << std::setprecision(20) << "tmp2 = " << tmp2 << std::endl;
    return 0;
}
```
the ouput is

```bash
tmp1 = 0.0099009899422526359558
tmp2 = 0.0099009900990099011103
```
 `auto tmp1 = static_cast<float>(0.009900990099009901);` This will cause tmp1 to become 0.0099009, resulting in a loss of precision, so the final result will not match the expected value.
I also found that the bug occurred at that position
https://github.com/pytorch/pytorch/blob/86d8af6a6cc648134289de89d393d0dce5b3a5f4/torch/_inductor/lowering.py#L6238

The commit states that the precision lost is expected in cuda implementation. 
original commit
https://github.com/pytorch/pytorch/commit/03439d4c1c39d6957d09b6024946c92849b30ffc
cuda implementation
https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

What is interesting is that the Triton kernel works correctly due to the precision of float type in python.
```python
def triton_poi_fused_div_0(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 25
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = 0.009900990099009901
    tmp2 = tmp0 * tmp1
    tl.store(out_ptr0 + (x0), tmp2, xmask)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

### CODE REVIEW

**Comment by thenumberouscode** [comment]
What is interesting is that the Triton kernel works correctly.
```python
def triton_poi_fused_div_0(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 25
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = 0.009900990099009901
    tmp2 = tmp0 * tmp1
    tl.store(out_ptr0 + (x0), tmp2, xmask)
```

**Comment by thenumberouscode** [comment]
@jgong5  Could you please take a look at my PR? Also, could you let me know who the reviewers are for those changes? Thank you!

**Comment by thenumberouscode** [comment]
@albanD Could you let me know who would be a good fit to review this PR?

**Review by eellison — COMMENTED** [review]
I think if we're going to disable this optimization we should do it on cuda as well. The divergence for this repro is `tensor(5.6025e-10, device='cuda:0')` for cuda and `tensor(-2.6776e-09)` for cpu.

@ngimel @jansel any thoughts on if this is within acceptable divergence ? multiplication is faster than division but most of these kernels should be bandwidth bound anyway so not sure it's worth the micro-optimizaiton.

**Comment by thenumberouscode** [comment]
> I think if we're going to disable this optimization we should do it on cuda as well. The divergence for this repro is `tensor(5.6025e-10, device='cuda:0')` for cuda and `tensor(-2.6776e-09)` for cpu.
> 
> @ngimel @jansel any thoughts on if this is within acceptable divergence ? multiplication is faster than division but most of these kernels should be bandwidth bound anyway so not sure it's worth the micro-optimizaiton.

@eellison I agree with your suggestion to remove the CUDA optimization as well and I’ve already changed the code. However, I’m not sure what you mean by 'divergence repro.' Could you please provide a complete demo code?

**Comment by ngimel** [comment]
cuda eager is actually using inverse multiplication optimization, so if we want to match results, we should be using different paths for cuda and cpu in inductor

**Comment by ngimel** [comment]
In addition, triton internally also replaces a lot of scalar divisions with inverse multiplications, so globally it's not clear what we want to achieve here

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -6229,14 +6229,6 @@ def div_prim(a, b):
     if is_integral:
         return truncdiv(a, b)
 
-    if (divisor := get_constant_value(b)) is not None:
```
Comment (eellison) [inline torch/_inductor/lowering.py:6220]: Okay, sorry, per @ngimel, let's do you your original approach and disable this opt only on cpu

Comment (thenumberouscode) [reply torch/_inductor/lowering.py:6220]: @eellison done, please review PR again.

**Review by eellison — APPROVED** [review]

**Comment by thenumberouscode** [comment]
@eellison The workflow failed due to a linting issue. I've made the necessary changes. Please run the workflow again. Thank you!

**Comment by thenumberouscode** [comment]
@eellison The workflow seems to have failed due to an unrelated test.
https://github.com/pytorch/pytorch/actions/runs/16460471120/job/46570705958?pr=158231
Could we run the workflow again? Additionally, could you grant me the privilege to run the workflow myself? I currently have five PRs under review, and I will be a full-time PyTorch developer in the next few months. Having access to run the workflow would save me a lot of time during the review process.
